### ISSUE TITLE
IndexOutOfBoundsException on Azure Benchmark

### ISSUE THREAD

**Issue by tggreene** [xtdb/crux#3537]
We're seeing an error after some time (~10 minutes) in a multi node setup across multiple nodes:

```
2024-07-30T11:03:25.434509749Z 11:03:25.434 [xtdb-tx-subscription-pool-1-thread-1] ERROR xtdb.util - Uncaught exception:
2024-07-30T11:03:25.434524757Z java.lang.IndexOutOfBoundsException: index: 844, length: 1 (expected: range(0, 0))
2024-07-30T11:03:25.434529346Z  at org.apache.arrow.memory.ArrowBuf.checkIndexD(ArrowBuf.java:299)
2024-07-30T11:03:25.434535127Z  at org.apache.arrow.memory.ArrowBuf.chk(ArrowBuf.java:285)
2024-07-30T11:03:25.434541318Z  at org.apache.arrow.memory.ArrowBuf.getByte(ArrowBuf.java:487)
2024-07-30T11:03:25.434547219Z  at org.apache.arrow.vector.BaseFixedWidthVector.isSet(BaseFixedWidthVector.java:831)
2024-07-30T11:03:25.434553310Z  at org.apache.arrow.vector.BaseFixedWidthVector.isNull(BaseFixedWidthVector.java:820)
2024-07-30T11:03:25.434559402Z  at org.apache.arrow.vector.BaseFixedWidthVector.copyFrom(BaseFixedWidthVector.java:887)
2024-07-30T11:03:25.434565583Z  at org.apache.arrow.vector.BaseFixedWidthVector.copyFromSafe(BaseFixedWidthVector.java:910)
2024-07-30T11:03:25.434569861Z  at xtdb.vector.ScalarVectorWriter.rowCopier$lambda$1(FieldVectorWriters.kt:63)
2024-07-30T11:03:25.434573618Z  at xtdb.vector.IndirectVectorReader.lambda$rowCopier$2(IndirectVectorReader.java:207)
2024-07-30T11:03:25.434577225Z  at xtdb.vector.IndirectMultiVectorReader.rowCopier$lambda$12(IndirectMultiVectorReader.kt:187)
2024-07-30T11:03:25.434590620Z  at xtdb.vector.IndirectVectorReader.lambda$rowCopier$2(IndirectVectorReader.java:207)
2024-07-30T11:03:25.434594397Z  at xtdb.vector.StructVectorWriter.rowCopier$lambda$20(StructVectorWriter.kt:195)
2024-07-30T11:03:25.434598033Z  at xtdb.vector.DenseUnionVectorWriter$ChildWriter.rowCopier$lambda$1(DenseUnionVectorWriter.kt:117)
2024-07-30T11:03:25.434601560Z  at xtdb.indexer$$reify__29554$fn__29571.invoke(indexer.clj:327)
2024-07-30T11:03:25.434605147Z  at xtdb.indexer.live_index.LiveTable$reify__21114.logPut(live_index.clj:132)
2024-07-30T11:03:25.434608633Z  at xtdb.indexer$$reify__29554.indexOp(indexer.clj:327)
2024-07-30T11:03:25.434612190Z  at xtdb.indexer$query_indexer$eval_query$reify__29609.accept(indexer.clj:399)
2024-07-30T11:03:25.434615746Z  at xtdb.operator.project.ProjectCursor$reify__24473.accept(project.clj:130)
2024-07-30T11:03:25.434619263Z  at xtdb.operator.project.ProjectCursor$reify__24473.accept(project.clj:130)
2024-07-30T11:03:25.434622810Z  at xtdb.operator.rename.RenameCursor$reify__25282.accept(rename.clj:39)
2024-07-30T11:03:25.434626386Z  at xtdb.operator.scan.TrieCursor$fn__23143.invoke(scan.clj:309)
2024-07-30T11:03:25.434630624Z  at xtdb.operator.scan.TrieCursor.tryAdvance(scan.clj:260)
2024-07-30T11:03:25.434636335Z  at xtdb.operator.rename.RenameCursor.tryAdvance(rename.clj:29)
2024-07-30T11:03:25.434642226Z  at xtdb.operator.project.ProjectCursor.tryAdvance(project.clj:116)
2024-07-30T11:03:25.434647265Z  at xtdb.operator.project.ProjectCursor.tryAdvance(project.clj:116)
2024-07-30T11:03:25.434652324Z  at xtdb.query$wrap_cursor$reify__29370.tryAdvance(query.clj:78)
2024-07-30T11:03:25.434658405Z  at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
2024-07-30T11:03:25.434662473Z  at xtdb.indexer$query_indexer$eval_query__29608.invoke(indexer.clj:396)
2024-07-30T11:03:25.434665990Z  at xtdb.indexer$wrap_sql_args$fn__29615.invoke(indexer.clj:430)
2024-07-30T11:03:25.434669526Z  at xtdb.indexer$foreach_arg_row.invokeStatic(indexer.clj:413)
2024-07-30T11:03:25.434673053Z  at xtdb.indexer$foreach_arg_row.invokePrim(indexer.clj)
2024-07-30T11:03:25.434676539Z  at xtdb.indexer$$reify__29620.indexOp(indexer.clj:401)
2024-07-30T11:03:25.434680106Z  at xtdb.indexer.Indexer$fn__29670$index_tx_ops__29673$fn__29686.invoke(indexer.clj:623)
2024-07-30T11:03:25.434683703Z  at clojure.lang.AFn.call(AFn.java:18)
2024-07-30T11:03:25.434687229Z  at io.micrometer.core.instrument.AbstractTimer.recordCallable(AbstractTimer.java:175)
2024-07-30T11:03:25.434690756Z  at xtdb.indexer.Indexer$fn__29670$index_tx_ops__29673.invoke(indexer.clj:620)
2024-07-30T11:03:25.434694272Z  at xtdb.indexer.Indexer$fn__29670$fn__29694.invoke(indexer.clj:634)
2024-07-30T11:03:25.434697769Z  at xtdb.indexer.Indexer$fn__29670.invoke(indexer.clj:633)
2024-07-30T11:03:25.434701275Z  at xtdb.indexer.Indexer.indexTx(indexer.clj:597)
2024-07-30T11:03:25.434704762Z  at xtdb.log.watcher$watch_log_BANG_$reify__30425.acceptRecord(watcher.clj:53)
2024-07-30T11:03:25.434708248Z  at xtdb.log$tx_handler$fn__21280.invoke(log.clj:37)
2024-07-30T11:03:25.434711885Z  at clojure.lang.PersistentVector.reduce(PersistentVector.java:343)
2024-07-30T11:03:25.434715422Z  at clojure.core$reduce.invokeStatic(core.clj:6885)
2024-07-30T11:03:25.434718948Z  at clojure.core$reduce.invoke(core.clj:6868)
2024-07-30T11:03:25.434722444Z  at xtdb.log$handle_polling_subscription$fn__21286.invoke(log.clj:53)
2024-07-30T11:03:25.434725921Z  at clojure.lang.AFn.run(AFn.java:22)
2024-07-30T11:03:25.434729408Z  at java.base/java.lang.Thread.run(Thread.java:1583)
```

**Comment by refset** [comment]
Possibly related to https://github.com/xtdb/xtdb/issues/2373 (IIOBE is similarly showing `length: 1 (expected: range(0, 0))`)

**Comment by danmason** [comment]
Wondering if this might be related to running eventhub log - probably worth running a cluster with just the disk cache & a local log to see what happens.

**Comment by tggreene** [comment]
Found this IOOBE issue again with event hub kafka with index 3157

**Comment by danmason** [comment]
Have continued to see this on numerous runs inside of Azure - most recent instance of this:

**Stack Trace:**
 ```
java.util.concurrent.ExecutionException: clojure.lang.ExceptionInfo: Ingestion stopped: Failed opening record batch 'tables/item/data/log-l01-fr4c870a-nr5113962-rs4b000.arrow' {}
 at java.util.concurrent.CompletableFuture.reportGet (CompletableFuture.java:396)
    java.util.concurrent.CompletableFuture.get (CompletableFuture.java:2073)
    clojure.core$deref_future.invokeStatic (core.clj:2317)
    clojure.core$deref.invokeStatic (core.clj:2337)
    clojure.core$deref.invoke (core.clj:2323)
    xtdb.indexer.Indexer.awaitTx (indexer.clj:568)
    xtdb.indexer$await_tx.invokeStatic (indexer.clj:714)
    xtdb.indexer$await_tx.invoke (indexer.clj:711)
    xtdb.bench.auctionmark$then_await_tx.invokeStatic (auctionmark.clj:623)
    xtdb.bench.auctionmark$then_await_tx.invoke (auctionmark.clj:621)
    xtdb.bench.auctionmark$benchmark$fn__30281.invoke (auctionmark.clj:718)
    xtdb.bench$compile_benchmark$run_benchmark__6110.invoke (bench.clj:305)
    xtdb.run_auctionmark.main$run_full_auctionmark$fn__30325.invoke (main.clj:92)
    xtdb.run_auctionmark.main$run_full_auctionmark.invokeStatic (main.clj:91)
    xtdb.run_auctionmark.main$run_full_auctionmark.invoke (main.clj:78)
Caused by: clojure.lang.ExceptionInfo: Ingestion stopped: Failed opening record batch 'tables/item/data/log-l01-fr4c870a-nr5113962-rs4b000.arrow'
{}
 at xtdb.await$__GT_ingester_ex.invokeStatic (await.clj:18)
    xtdb.await$__GT_ingester_ex.invoke (await.clj:17)
    xtdb.await$await_tx_async$fn__20931.invoke (await.clj:31)
    xtdb.await$await_tx_async.invokeStatic (await.clj:26)
    xtdb.await$await_tx_async.invoke (await.clj:22)
    xtdb.indexer.Indexer.awaitTx (indexer.clj:679)
    xtdb.indexer$await_tx.invokeStatic (indexer.clj:714)
    xtdb.indexer$await_tx.invoke (indexer.clj:711)
    xtdb.bench.auctionmark$then_await_tx.invokeStatic (auctionmark.clj:623)
    xtdb.bench.auctionmark$then_await_tx.invoke (auctionmark.clj:621)
    xtdb.bench.auctionmark$benchmark$fn__30281.invoke (auctionmark.clj:718)
    xtdb.bench$compile_benchmark$run_benchmark__6110.invoke (bench.clj:305)
    xtdb.run_auctionmark.main$run_full_auctionmark$fn__30325.invoke (main.clj:92)
    xtdb.run_auctionmark.main$run_full_auctionmark.invokeStatic (main.clj:91)
    xtdb.run_auctionmark.main$run_full_auctionmark.invoke (main.clj:78)
Caused by: clojure.lang.ExceptionInfo: Failed opening record batch 'tables/item/data/log-l01-fr4c870a-nr5113962-rs4b000.arrow'
{:path #object[sun.nio.fs.UnixPath 0x7404d320 "tables/item/data/log-l01-fr4c870a-nr5113962-rs4b000.arrow"], :block-idx 498}
 at xtdb.buffer_pool$open_record_batch$fn__8625.invoke (buffer_pool.clj:525)
    xtdb.buffer_pool$open_record_batch.invokeStatic (buffer_pool.clj:516)
    xtdb.buffer_pool$open_record_batch.invoke (buffer_pool.clj:515)
    xtdb.operator.scan.ArrowMergePlanPage.load_page (scan.clj:357)
    xtdb.operator.scan.TrieCursor$fn__23128$iter__23134__23138$fn__23139$fn__23140.invoke (scan.clj:258)
    xtdb.operator.scan.TrieCursor$fn__23128$iter__23134__23138$fn__23139.invoke (scan.clj:257)
    clojure.lang.LazySeq.force (LazySeq.java:50)
    clojure.lang.LazySeq.realize (LazySeq.java:89)
    clojure.lang.LazySeq.seq (LazySeq.java:106)
    clojure.lang.RT.seq (RT.java:555)
    clojure.core$seq__5484.invokeStatic (core.clj:139)
    clojure.core.protocols$seq_reduce.invokeStatic (protocols.clj:24)
    clojure.core.protocols$fn__8258.invokeStatic (protocols.clj:74)
    clojure.core.protocols/fn (protocols.clj:74)
    clojure.core.protocols$fn__8199$G__8194__8212.invoke (protocols.clj:13)
Caused by: java.lang.IndexOutOfBoundsException: index: -6, length: 6 (expected: range(0, 0))
 at org.apache.arrow.memory.ArrowBuf.checkIndexD (ArrowBuf.java:299)
    org.apache.arrow.memory.ArrowBuf.checkBytes (ArrowBuf.java:112)
[13:33](https://juxt.slack.com/archives/C077A221N4V/p1724243614992719)
org.apache.arrow.memory.util.ByteFunctionHelpers.compare (ByteFunctionHelpers.java:220)
    xtdb.util$validate_arrow_magic.invokeStatic (util.clj:388)
    xtdb.util$validate_arrow_magic.invoke (util.clj:387)
    xtdb.util$read_arrow_footer.invokeStatic (util.clj:398)
    xtdb.util$read_arrow_footer.invoke (util.clj:397)
    xtdb.buffer_pool$open_record_batch$fn__8625.invoke (buffer_pool.clj:518)
    xtdb.buffer_pool$open_record_batch.invokeStatic (buffer_pool.clj:516)
    xtdb.buffer_pool$open_record_batch.invoke (buffer_pool.clj:515)
    xtdb.operator.scan.ArrowMergePlanPage.load_page (scan.clj:357)
    xtdb.operator.scan.TrieCursor$fn__23128$iter__23134__23138$fn__23139$fn__23140.invoke (scan.clj:258)
    xtdb.operator.scan.TrieCursor$fn__23128$iter__23134__23138$fn__23139.invoke (scan.clj:257)
    clojure.lang.LazySeq.force (LazySeq.java:50)
    clojure.lang.LazySeq.realize (LazySeq.java:89)
```

Thrown by two separate nodes at around the same time (during the `then-await` portion of the run, but seemingly also thrown earlier in the OLTP loop when trying to proc new items/prepare a new query.

**Comment by danmason** [comment]
Examining the file in question, `log-l01-fr4c870a-nr5113962-rs4b000.arrow`, I noted the following:
- The file exists in the blob container.
- The file is of size zero bytes.

Fairly strange indeed. I decided to take a look at some of the earlier logs of the container in the multi node config.

Looking at the logs of the first node in the cluster:
- I see a long host of these IOOBE errors when attempting to call any query/transaction against items. Always `Ingestion stopped: Failed opening record batch 'tables/item/data/log-l01-fr4c870a-nr5113962-rs4b000.arrow' {}`
- Saw a few messages attempting and failing to write a multipart upload to 'log-l01-fr4c870a-nr5113962-rs4b000.arrow', `INFO  xtdb.azure.object-store - Blob already exists for https://xtdbazurebenchmark.blob.core.windows.net/xtdbazurebenchmarkcontainer/xtdb-object-store%2Fv04%2Ftables%2Fitem%2Fdata%2Flog-l01-fr4c870a-nr5113962-rs4b000.arrow - aborting multipart upload`

Taking a look at the logs of the second node in the cluster:
- Much of the same errors as the above.

Finally, the third node:
- One instance of failing to write the file and aborting multipart upload.
- A number of outOfMemory and outOfDirectMemory errors within compaction job.
  - Not sure if related but definitely want to investigate this seperately - shall raise a card.

Now, in terms of how we are running multi node benchmarks, it current looks like this:
- We run an initcontainer that does the initial load phase of the node on its own.
- We runs three containers/nodes in parallel afterwards for an hour.

As such, decided to look at the logs from the (seemingly successful) load phase, some notes from that:
- Logs for main stages seem okay.
- We've seen a fair amount of `interrupted` exceptions when closing this node, thrown mid multipart upload when the node closes.
- Could be related to the above?

Error in question - doesn't necessarily tell us a lot but wondering if cleaner shutdowns could prevent this issue:
```
11:31:26.484 [xtdb.compactor-22] ERROR xtdb.compactor - Error running compaction job.
reactor.core.Exceptions$ReactiveException: java.lang.InterruptedException
	at reactor.core.Exceptions.propagate(Exceptions.java:396)
	at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:92)
	at reactor.core.publisher.Mono.block(Mono.java:1742)
	at com.azure.storage.common.implementation.StorageImplUtils.blockWithOptionalTimeout(StorageImplUtils.java:146)
	at com.azure.storage.blob.specialized.BlockBlobClient.stageBlockWithResponse(BlockBlobClient.java:787)
	at com.azure.storage.blob.specialized.BlockBlobClient.stageBlock(BlockBlobClient.java:704)
	at xtdb.azure.object_store.MultipartUpload.uploadPart(object_store.clj:84)
	at xtdb.buffer_pool$upload_multipart_buffers.invokeStatic(buffer_pool.clj:264)
	at xtdb.buffer_pool$upload_multipart_buffers.invoke(buffer_pool.clj:254)
	at xtdb.buffer_pool$upload_arrow_file$fn__8446.invoke(buffer_pool.clj:313)
	at xtdb.buffer_pool$upload_arrow_file.invokeStatic(buffer_pool.clj:312)
	at xtdb.buffer_pool$upload_arrow_file.invoke(buffer_pool.clj:306)
	at xtdb.buffer_pool.RemoteBufferPool$reify__8505.end(buffer_pool.clj:386)
	at xtdb.trie$open_trie_writer$reify__15735.end(trie.clj:195)
	at xtdb.Compactor.writeRelation(Compactor.kt:193)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__15995$fn__15996$fn__16005$fn__16006$fn__16007.invoke(compactor.clj:144)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__15995$fn__15996$fn__16005$fn__16006.invoke(compactor.clj:141)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__15995$fn__15996$fn__16005.invoke(compactor.clj:137)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__15995$fn__15996.invoke(compactor.clj:137)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__15995.invoke(compactor.clj:126)
	at xtdb.compactor$exec_compaction_job_BANG_.invokeStatic(compactor.clj:126)
	at xtdb.compactor$exec_compaction_job_BANG_.invoke(compactor.clj:121)
	at xtdb.compactor$eval16111$fn__16113$compactor_task__16142$fn__16144.invoke(compactor.clj:275)
	at xtdb.compactor$eval16111$fn__16113$compactor_task__16142.invoke(compactor.clj:273)
	at xtdb.compactor$eval16111$fn__16113$fn__16147$fn__16148$fn__16149$fn__16155.invoke(compactor.clj:310)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.lang.VirtualThread.run(VirtualThread.java:329)
	Suppressed: reactor.core.Exceptions$ReactiveException: java.lang.InterruptedException
		at reactor.core.Exceptions.propagate(Exceptions.java:396)
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:92)
		at reactor.core.publisher.Mono.block(Mono.java:1742)
		at com.azure.storage.common.implementation.StorageImplUtils.blockWithOptionalTimeout(StorageImplUtils.java:146)
		at com.azure.storage.blob.specialized.BlockBlobClient.commitBlockListWithResponse(BlockBlobClient.java:1146)
		at com.azure.storage.blob.specialized.BlockBlobClient.commitBlockListWithResponse(BlockBlobClient.java:1094)
		at com.azure.storage.blob.specialized.BlockBlobClient.commitBlockList(BlockBlobClient.java:1044)
		at com.azure.storage.blob.specialized.BlockBlobClient.commitBlockList(BlockBlobClient.java:1013)
		at xtdb.azure.object_store.MultipartUpload$fn__30713.invoke(object_store.clj:104)
		at xtdb.azure.object_store.MultipartUpload.abort(object_store.clj:102)
		at xtdb.buffer_pool$upload_multipart_buffers$fn__8433.invoke(buffer_pool.clj:274)
		at xtdb.buffer_pool$upload_multipart_buffers.invokeStatic(buffer_pool.clj:273)
		... 21 common frames omitted
```

**Comment by danmason** [comment]
Taking a look at it, the timings line up that suggest it could well be the non graceful shutdown of the compactor causing issues, in particular, the timestamps involved. Our 0 byte file, `log-l01-fr4c870a-nr5113962-rs4b000.arrow`, was written at 12:31:26 PM.

The logs from our interrupted compactor job?
```
11:31:26.484 [xtdb.compactor-22] ERROR xtdb.compactor - Error running compaction job.
```

Given that this is fairly specific to our multinode setup, I'd say its unlikely the original issue with IOOBEs on Azure was this problem, but seems to be the cause of our current problems.

**Comment by danmason** [comment]
Fixed by #3629
